[5.4] Allow opt-out from automated updates during installation

SniperSister opened the pull request:
### Summary of Changes
This PR adds an additional section to the installation process, allowing users to opt out from automated updates. Using that option will prevent a site from registering with the service in the first place, ensuring that no data is submitted.


### Testing Instructions
* Install a new site without the fix, e.g. by using the latest nightly build: https://developer.joomla.org/nightlies/Joomla_5.4.0-alpha4-dev-Development-Full_Package.zip
* After installation, the automated update feature is enabled
* Install a new site with the fix by using the package for this https://artifacts.joomla.org/drone/joomla/joomla-cms/5.4-dev/45697/downloads/87039/Joomla_5.4.0-alpha4-dev+pr.45697-Development-Full_Package.zip
* During installation, opt-out from auto-updates using the respective button in the installation finalisation screen
* Log in into the administration: automated updates are disabled


### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [ ] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

SniperSister committed:
Allow install optout

SniperSister committed:
cs fix

brianteeman commented on installation/language/en-GB/joomla.ini:
```diff
@@ -144,6 +144,13 @@ INSTL_LANGUAGES_WARNING_NO_INTERNET="Joomla was unable to connect to the languag
 
+; Automated Updates Opt Out
+INSTL_AUTOMATED_UPDATES="Automated Updates"
+INSTL_AUTOMATED_UPDATES_DESC="Joomla has a built-in mechanism that will update your Joomla installation automatically.<br /><br /><strong>Please note:</strong> enabling Automated Updates will register your site at the respective service provided by the Joomla! project. The registration allows the Joomla project to access information about your site and server environment, specifically Site Url, P ...
```
we dont use xhtml ;)

```suggestion
INSTL_AUTOMATED_UPDATES_DESC="Joomla has a built-in mechanism that will update your Joomla installation automatically.<br><br><strong>Please note:</strong> enabling Automated Updates will register your site at the respective service provided by the Joomla! project. The registration allows the Joomla project to access information about your site and server environment, specifically Site Url, PHP version, Database type and version, CMS version and Server OS type and version. This information is only used to improve the service.<br><br>Alternatively you can disable automated updates using the button below."
```

brianteeman commented on installation/src/Controller/InstallationController.php:
```diff
@@ -299,4 +299,44 @@ public function delete()
     }
+
+    /**
+     * Opt out from automated updartes
```
```suggestion
     * Opt out from automated updates
```

brianteeman commented on installation/src/Controller/InstallationController.php:
```diff
@@ -299,4 +299,44 @@ public function delete()
+        $model = $this->getModel('AutomatedUpdates');
+
+        if (!$model->disable()) {
+            // We can't send a response with sendJsonResponse because our installation classes might not now exist
```
```suggestion
            // We can't send a response with sendJsonResponse because our installation classes might not exist yet
```

brianteeman commented on installation/src/Controller/InstallationController.php:
```diff
@@ -299,4 +299,44 @@ public function delete()
+
+        $this->app->getSession()->destroy();
+
+        // We can't send a response with sendJsonResponse because our installation classes now do not exist
```
```suggestion
        // We can't send a response with sendJsonResponse because our installation classes do not exist yet
```

assuming i understand correctly

SniperSister committed:
Apply suggestions from code review

Co-authored-by: Brian Teeman <brian@teeman.net>

brianteeman commented on installation/language/en-GB/joomla.ini:
```diff
@@ -144,6 +144,13 @@ INSTL_LANGUAGES_WARNING_NO_INTERNET="Joomla was unable to connect to the languag
 
+; Automated Updates Opt Out
+INSTL_AUTOMATED_UPDATES="Automated Updates"
+INSTL_AUTOMATED_UPDATES_DESC="Joomla has a built-in mechanism that will update your Joomla installation automatically.<br><br><strong>Please note:</strong> enabling Automated Updates will register your site at the respective service provided by the Joomla! project. The registration allows the Joomla project to access information about your site and server environment, specifically Site Url, PHP v ...
```
![image](https://github.com/user-attachments/assets/56a7d522-27e2-4433-9664-3c20ec90db63)

Isn't this just for minor and path releases? If so then perhaps the first sentence should say something so that people dont expect automated updates forever.

They are enabled by default so saying "enabling automated updates..." is not correct and you can just delete the word "enabling"

SniperSister replied:
> They are enabled by default so saying "enabling automated updates..." is not correct and you can just delete the word "enabling"

Fixed! 

> Isn't this just for minor and path releases? If so then perhaps the first sentence should say something so that people dont expect automated updates forever.

I see your point, I'm just worried that the paragraph becomes even longer with additional information. Any thoughts?

brianteeman replied:
Perhaps

"Joomla automatically updates itself for minor and patch releases."

muhme replied:
@SniperSister Can you possibly just resolve this valid comment with the suggestion? We would like to test and merge this feature as soon as possible.

SniperSister replied:
Done

brianteeman commented on installation/language/en-GB/joomla.ini:
```diff
@@ -144,6 +144,13 @@ INSTL_LANGUAGES_WARNING_NO_INTERNET="Joomla was unable to connect to the languag
+; Automated Updates Opt Out
+INSTL_AUTOMATED_UPDATES="Automated Updates"
+INSTL_AUTOMATED_UPDATES_DESC="Joomla has a built-in mechanism that will update your Joomla installation automatically.<br><br><strong>Please note:</strong> enabling Automated Updates will register your site at the respective service provided by the Joomla! project. The registration allows the Joomla project to access information about your site and server environment, specifically Site Url, PHP v ...
+INSTL_AUTOMATED_UPDATES_DISABLE="Disable automated updates"
```
```suggestion
INSTL_AUTOMATED_UPDATES_DISABLE="Disable Automated Updates"
```

follow the UI style guide

SniperSister committed:
Update installation/language/en-GB/joomla.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

SniperSister committed:
update opt-out text

SniperSister committed:
Merge branch '54-autoupdate-installer' of https://github.com/SniperSister/joomla-cms into 54-autoupdate-installer

muhme committed:
Merge branch '5.4-dev' into 54-autoupdate-installer

SniperSister committed:
Update i18n string

SniperSister committed:
Merge branch '54-autoupdate-installer' of https://github.com/SniperSister/joomla-cms into 54-autoupdate-installer

ceford commented:
I think the test instructions need some clarification. What I did seemed to work first time but I needed to check and what I did the second time did not work (the Autoupdates was still enabled). My procedure:

1. Installed latest Nightly Build full package for 5.4.0-alpha3-dev.
2. Installed Patchtester and applied patch.
3. Installed latest Nightly Build update package for 5.4.0-alpha3-dev.
4. Back to Home Dashboard to see status of Automated Updates.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/45697">issues.joomla.org/tracker/joomla-cms/45697</a>.</sub>

SniperSister commented:
@ceford thank you so much for taking the time to test the patch, it's very much appreciated!

You are right, the test instructions have been unclear! I've updated the description and provided some more meaningful instructions.

ceford commented:
Tested with new instructions. 

With patch 45696 the new installation seemed to be as expected. Auto-update was not enabled on installation. 

With patch 45697 and Disable Automated Updates not selected the installation was normal. However:

With patch 45697 and Disable Automated Updates selected there was a snag: the Automated Updates box disappeared but the Install Additional Languages button and the Remove Installation Folder button were both greyed. The Open Administrator button worked but I can see that the installation folder is still present.
<img width="1440" height="1074" alt="image" src="https://github.com/user-attachments/assets/e7f71aa6-4f5b-4490-88b3-a70abdcfcca4" />
<img width="1440" height="687" alt="image" src="https://github.com/user-attachments/assets/fa2c4dc7-e3f5-4d88-81df-f1f369ebc08a" />

muhme committed:
Merge branch '5.4-dev' into 54-autoupdate-installer

SniperSister committed:
fix stan error

muhme commented:
I have tested this item :white_check_mark: successfully on 27dfd625b8f218a7d66606e79cd6750d3fec4aed<br /><br />Beforehand: I could not reproduce the issue with the disabled 'Remove installation folder' button. I removed the installation folder each time.  

Tested with manual installation from full package on **local site**
* 1st time disabled Automated Updates on remove view and checked Automated Updates are disabled
* 2nd time didn't disable Automated Updates on remove view and checked Automated Updates:
  * Home Dashboard call there is yellow 'Automated Updates unavailable.'
  * Click in 'Automated Update' shows Automated Update: No
  * I can enable Automated Update and Save & Close, afterwards error message
    * The automated update service is unavailable for your site as it's not accessible from the internet. Automated updates have been disabled.
    * and Automated Updates remains disabled

Tested with manual installation from full package on **public site**
* Didn't disable Automated Updates on remove view and checked Automated Updates:
  * ✅ Home Dashboard shows green 'Automated Updates are enabled.'
  * ✅ Click in 'Automated Update' shows Automated Update: Yes, update token and last checked date today<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/45697">issues.joomla.org/tracker/joomla-cms/45697</a>.</sub>

SniperSister commented:
> I have updated the branch to get progress on this PR, to be included within beta1, but PHPStan fails with Instantiated class JConfig not found. Please assist.

Fixed

richard67 commented:
I've updated the links to packages in the description of this PR so testers can be sure to have packages which are up to date with latest 5.4-dev.

richard67 committed:
Merge branch '5.4-dev' into 54-autoupdate-installer

richard67 commented:
P.S.: I think that it is here where the buttons are disabled after deleting the installation folder: https://github.com/joomla/joomla-cms/pull/45697/files#diff-d9f8a6c89739f0469e6ad48530ab28a41675604c9aad1d871c83a5f3f8032285R109-R121

SniperSister committed:
Disable optout button on install folder removal

SniperSister committed:
Merge branch '54-autoupdate-installer' of https://github.com/SniperSister/joomla-cms into 54-autoupdate-installer

richard67 committed:
Merge branch '5.4-dev' into 54-autoupdate-installer

richard67 commented:
@ceford The disabled button issue should be solved now with the latest changes.

SniperSister committed:
not only disable but removce section

SniperSister committed:
Merge branch '54-autoupdate-installer' of https://github.com/SniperSister/joomla-cms into 54-autoupdate-installer

richard67 commented on installation/template/js/remove.js:
```diff
@@ -41,6 +41,17 @@ if (document.getElementById('removeInstallationFolder')) {
+  document.getElementById('automatedUpdatesDisableButton')
+    .addEventListener('click', function (e) {
+      e.preventDefault();
+      let confirm = window.confirm(Joomla.Text._('INSTL_DISABLE_AUTOUPDATE').replace('%s', 'installation'));
```
I don't think it needs the `.replace('%s', 'installation')` here as that language string does not have any placeholder for the name of the installation folder.

SniperSister committed:
remove unused statemtent

richard67 commented:
I have tested this item :white_check_mark: successfully on e23dc6fa78da15a11d68bc0991a32a6b41dcc65d<br /><br />I've also tested this additional test case:

At the final page, directly use the button to remove the installation folder.

Result: Anything related to the languages installation, the new option added by this PR here and the button to remove the installation folder are removed from the page because all that stuff cannot be used after the installation folder has been removed, and only the buttons for going to administrator or site remain.

This is the existing behaviour without this PR. The PR does not change anything on it, it only makes sure the new option is also handled like this.

Maybe not good UX, but nothing to be fixed with this PR here.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/45697">issues.joomla.org/tracker/joomla-cms/45697</a>.</sub>

richard67 commented:
@ceford Is is a Windows server? If so, this is due to issue #37909 . That is the expected behaviour for that case. If everything else was working for you, I would call it a successful test. But if it is Linux then there might be a permission issue when deleting the installation folder. That would have the same result.

SniperSister committed:
remove unused session-related statement

richard67 committed:
Merge branch '5.4-dev' into 54-autoupdate-installer

SniperSister committed:
catch sequential click of both buttons

SniperSister committed:
Merge branch '54-autoupdate-installer' of https://github.com/SniperSister/joomla-cms into 54-autoupdate-installer

SniperSister commented:
@richard67 took a closer look at the button problem and suggested a bunch of changes, thank you!

Please retry.

richard67 commented:
@ceford Thanks a lot for your effort for testing.

richard67 commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/45697">issues.joomla.org/tracker/joomla-cms/45697</a>.</sub>

muhme committed:
Merge branch '5.4-dev' into 54-autoupdate-installer

muhme commented:
✅ Final test before merge with multiple manual full package installation on public site (always disabling AU afterwards)
* English (en-GB) Not disabled AU, delete installation folder, opened administrator -> AU is enabled, token is there
* German (Germany) Disable AU, delete installation folder is working with removed installation and disabled AU
* English (en-US) Not disable AU, not delete installation folder, opened administrator -> AU is enabled, token is there, installation folder exists
* Spanish (es-ES) First delete installation folder, no button Disable AU any more, AU is enabled, token is there, installation folder is deleted

muhme commented:
Thank you @SniperSister and @brianteeman for your contribution. Thank you @ceford and @richard67 for repeated testing.
